gcloud/rest/auth/token.py

```diff
@@ -1,95 +1,199 @@
+"""Google Cloud auth via service account file"""
 import datetime
+import enum
 import json
-import logging
 import os
 import threading
 import time
+from typing import Any
+from typing import Dict
+from typing import List  # pylint: disable=unused-import
+from typing import Optional
+
 try:
     from urllib.parse import urlencode
+    from urllib.parse import quote_plus
 except ImportError:
     from urllib import urlencode
+    from urllib import quote_plus
 
-# N.B. the cryptography library is required when calling jwt.encrypt() with
-# algorithm='RS256'. It does not need to be imported here, but this allows us
-# to throw this error at load time rather than lazily during normal operations,
-# where plumbing this error through will require several changes to otherwise-
-# good error handling.
-import cryptography  # pylint: disable=unused-import
+import backoff
 import jwt
 import requests
 
 
-TOKEN_URI = 'https://accounts.google.com/o/oauth2/token'
-TIMEOUT = 60
-
-log = logging.getLogger(__name__)
+GCE_METADATA_BASE = 'http://metadata.google.internal/computeMetadata/v1'
+GCE_METADATA_HEADERS = {'metadata-flavor': 'Google'}
+GCE_ENDPOINT_PROJECT = '{}/project/project-id'.format(GCE_METADATA_BASE)
+GCE_ENDPOINT_TOKEN = \
+    '{}/instance/service-accounts/default/token?recursive=true'\
+    .format(GCE_METADATA_BASE)
+GCLOUD_TOKEN_DURATION = 3600
+REFRESH_HEADERS = {'Content-Type': 'application/x-www-form-urlencoded'}
+
+
+class Type(enum.Enum):
+    AUTHORIZED_USER = 'authorized_user'
+    GCE_METADATA = 'gce_metadata'
+    SERVICE_ACCOUNT = 'service_account'
+
+
+def get_service_data(service):
+    # type: (Optional[str]) -> Dict[str, Any]
+    service = service or os.environ.get('GOOGLE_APPLICATION_CREDENTIALS')
+    if not service:
+        cloudsdk_config = os.environ.get('CLOUDSDK_CONFIG')
+        sdkpath = (cloudsdk_config
+                   or os.path.join(os.path.expanduser('~'), '.config',
+                                   'gcloud'))
+        service = os.path.join(sdkpath, 'application_default_credentials.json')
+        set_explicitly = bool(cloudsdk_config)
+    else:
+        set_explicitly = True
+
+    try:
+        with open(service, 'r') as f:
+            data = json.loads(f.read())
+            return data
+    except FileNotFoundError:
+        # only warn users if they have explicitly set the service_file path
+        if set_explicitly:
+            raise
+        return {}
+    except Exception:  # pylint: disable=broad-except
+        return {}
 
 
 class Token(object):
-    def __init__(self, creds=None, google_api_lock=None, scopes=None,
-                 timeout=TIMEOUT):
-        self.creds = creds or os.getenv('GOOGLE_APPLICATION_CREDENTIALS')
-        if not self.creds:
-            raise Exception('could not load service credentials')
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self,
+                 service_file=None,  # type: Optional[str]
```

Comment:
Let's accept the old arguments and ordering of args for legacy reasons along with a `DeprecationWarning` so we can do a deprecation period rather than a hard cutover.

Reply:
From what I understood of this comment, are you saying that the Token __init__ look like what we had previously, but the new args like service_file, session added in the end? And then if creds or timeout is provided -> fire the deprecation warning? not quite sure what the expectation is here. I think we need to keep the new args as well for cases like - https://github.com/talkiq/gcloud-aio/blob/master/datastore/gcloud/aio/datastore/datastore.py#L61 which we don't want to fail when this PR is merged. What do you think?

Reply:
Right! My thought would be something like:
```
def __init__(self, old_arg_1=None, old_arg_2=None, new_arg_1=None, new_arg_2=None, ...):
    if old_arg_1:
        warnings.warn('old_arg_1 is deprecated, use new_arg_1', DeprecationWarning)
        new_arg_1 = old_arg_1
    ... etc
```